Fix differences between php and php-fpm for scriptname

wpoely86 opened the pull request:
Originally done by @niobos in #7719. If you want to run owncloud with
php-fpm on RHEL 7, you need this. There is a bug in php that causes
$_SERVER['SCRIPT_NAME'] to also contains the PATH_INFO. This works
around that issue.

I need this patch to get owncloud to work on RHEL7 with apache 2.4 and php-fpm.

wpoely86 committed:
Fix differences between php and php-fpm for scriptname

Originally done by @niobos in #7719. If you want to run owncloud with
php-fpm on RHEL 7, you need this. There is a bug in php that causes
$_SERVER['SCRIPT_NAME'] to also contains the PATH_INFO. This works
around that issue.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/2e1dd9d9-39f8-46a3-961e-4598ed515a92) was created.

wpoely86 commented:
This contribution is MIT licensed.

karlitschek commented:
@DeepDiver1975 @LukasReschke What do you think?

DeepDiver1975 commented:
hmmm - this can only be accepted with additional unit tests.

Furthermore I lack knowledge on fpm - I cannot really judge.

@wpoely86 you state this is an issue with php - which version? Is there some bug report upstream? Thanks a lot!

wpoely86 commented:
@DeepDiver1975 please read #7719. It's all explained there.

DeepDiver1975 commented:
> @DeepDiver1975 please read #7719. It's all explained there.

THX - I missed that part. TBH: I'm undecided if we should merge this one in.

wpoely86 commented:
Well, the bugs are all fixed upstream, but the fixes are not backported to RHEL 7 (and its derivatives). Currently, owncloud with php-fpm is not usable on RHEL 7. This patch works around that. I agree that it is not 'nice'. Currently, I have to apply this patch manually everything owncloud gets updated...

th3fallen commented on lib/private/appframework/http/request.php:
```diff
@@ -522,8 +522,18 @@ public function getServerProtocol() {
+			} else {
+				$script_name = substr($_SERVER['SCRIPT_NAME'], 0, $pos);
+			}
+		}
```
please introduce a new method for this to cut down on the multiple cut and pastes

th3fallen commented on lib/private/appframework/http/request.php:
```diff
@@ -548,6 +558,15 @@ public function getRawPathInfo() {
+			} else {
+				$scriptName = substr($_SERVER['SCRIPT_NAME'], 0, $pos);
+			}
+		}
```
same here

th3fallen commented on lib/private/appframework/http/request.php:
```diff
@@ -604,6 +623,14 @@ public function getPathInfo() {
+			} else {
+				$name = substr($_SERVER['SCRIPT_NAME'], 0, $pos);
+			}
+		}
```
same here

Xenopathic commented:
Since the original code was posted by @niobos, I *think* they hold copyright over this change (but any modifications made by you in response to comments are your copyright). @niobos, could you quickly post here stating the license of this contribution?

Xenopathic commented on lib/private/appframework/http/request.php:
```diff
@@ -522,8 +522,18 @@ public function getServerProtocol() {
 		$uri = isset($this->server['REQUEST_URI']) ? $this->server['REQUEST_URI'] : '';
+
+		$script_name = $_SERVER['SCRIPT_NAME'];
+		if (array_key_exists('PATH_INFO', $_SERVER) === true) {
```
No need to explicitly check for trueness here, `array_key_exists()` only ever returns a bool.

Xenopathic commented on lib/private/appframework/http/request.php:
```diff
@@ -522,8 +522,18 @@ public function getServerProtocol() {
+		if (array_key_exists('PATH_INFO', $_SERVER) === true) {
+			$pos = strpos($_SERVER['SCRIPT_NAME'], rawurldecode($_SERVER['PATH_INFO']));
+			if ($pos === false) {
+			} else {
```
Avoid the empty `if` block, just use `if ($pos !== false)`

niobos commented:
I was monitoring this thread to see if anyone would notice that @wpoely86 licensed code that he does not hold copyright for. I was going to make a stand out of principle, so I am very happy to see that @Xenopathic did notice. Thank you!

I don't consider my patch to be a "creative work", so I don't even think copyright applies, but anyway: I hereby license my patch under the MIT license.

wpoely86 commented:
@niobos It's not a direct copy & paste of your patch. I've adapted it to 8.1 which means it's derived work, I think? I don't want to take any credit for the patch. I just want this fixed. Feel free to open a PR of your own and I will close this one.

wpoely86 committed:
Added getRawScriptName() to avoid code duplication

The fix for the difference between php and php-fpm for the scriptname is
now done in one place.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/a9d1780a-ef0a-4d12-b650-666fb1e9da02) was created.

niobos commented:
@wpoely86 Yes, in that case it's derived work, but that doesn't change the fact that you need my permission. Either way, I've granted you (and everyone else) that permission, so this discussion is pointless.

Thank you for pursuing this and doing this PR and all requested patches, I also want this fixed. I appreciate the effort!

wpoely86 commented:
@niobos I just want to make clear that I meant no harm or disrespect for your work.

Xenopathic commented:
:+1:

LukasReschke commented on lib/private/appframework/http/request.php:
```diff
@@ -522,12 +522,32 @@ public function getServerProtocol() {
+         * @return string
+         */
+        private function getRawScriptName() {
+            $script_name = $_SERVER['SCRIPT_NAME'];
```
Use $this->server

LukasReschke replied:
Or explain why you can't also below in 540 as others will surely adjust it otherwise

LukasReschke commented on lib/private/appframework/http/request.php:
```diff
@@ -522,12 +522,32 @@ public function getServerProtocol() {
 
+        /**
+         * Returns the actual scriptname. In some version of php there is a bug
+         * which causes PATH_INFO to also be part of SCRIPT_NAME.
```
Describe which versions. Add a link to the issue.

LukasReschke commented on lib/private/appframework/http/request.php:
```diff
@@ -522,12 +522,32 @@ public function getServerProtocol() {
+                }
+            }
+
+            return $script_name;
```
Do use camel-case such as `scriptName`

LukasReschke commented on lib/private/appframework/http/request.php:
```diff
@@ -522,12 +522,32 @@ public function getServerProtocol() {
+        private function getRawScriptName() {
+            $script_name = $_SERVER['SCRIPT_NAME'];
+
+            if (array_key_exists('PATH_INFO', $_SERVER) == true) {
```
Always use strict comparisons with `===`

LukasReschke commented:
Just my two cents: I'd more appreciate that somebody bumps the bug at RedHat as this is a bug as per http://php.net/ChangeLog-5.php#5.5.18:

> Fixed bug #65641 (PHP-FPM incorrectly defines the SCRIPT_NAME variable when using Apache, mod_proxy-fcgi and ProxyPass).

I'd really like to avoid adding this hackery. Really as in: Really really really ;-)

LukasReschke commented:
Reasons against adding strange workarounds:

1. We do not have a test environment for this. We did not notice this on our own. I therefore consider this not a primarily supported environment.
2. This change is not unit-tested.
3. This certainly is an upstream bug.
4. In some time nobody will understand what this is for and this can affect other code as well.

Did anybody file a bug upstream at RedHat? If so what is the issue reference number?

From my stance: If there is a distribution not properly backporting bug fixes we should not start adding ugly workarounds. That said we can also add a big red warning again at the installer if we encounter this bug (we need to come up with an heuristic though) to force RHEL users to complain loud enough ;-)
(it worked with the Null-Byte as well :speak_no_evil:)

LukasReschke commented on lib/private/appframework/http/request.php:
```diff
@@ -547,7 +567,7 @@ public function getRawPathInfo() {
 		}
 
-		$scriptName = $this->server['SCRIPT_NAME'];
+		$scriptName = getRawScriptName();
```
$this->getRawScriptName?

LukasReschke commented on lib/private/appframework/http/request.php:
```diff
@@ -603,7 +623,8 @@ public function getPathInfo() {
 	 */
 	public function getScriptName() {
-		$name = $this->server['SCRIPT_NAME'];
+		$name = getRawScriptName();
```
$this->getRawScriptName?

LukasReschke commented on lib/private/appframework/http/request.php:
```diff
@@ -522,12 +522,32 @@ public function getServerProtocol() {
+
 		if($this->config->getSystemValue('overwritewebroot') !== '' && $this->isOverwriteCondition()) {
-			$uri = $this->getScriptName() . substr($uri, strlen($this->server['SCRIPT_NAME']));
+			$uri = $this->getScriptName() . substr($uri, strlen(getRawScriptName()));
```
$this->getRawScriptName?

danmac-uk commented:
Exactly the same problem is also evident on Ubuntu 14.04, as documented in the previously posted issue, since ownCloud 6 as far as I'm aware. It would be real nice to be able to hit upgrade and not have to deal with this bug in a slightly different manner each time :(

danmac-uk commented:
This one works fine for me:
https://github.com/wpoely86/core/commit/5a49df542697d52bfbde5994ffcfeba88fc8ab9a.patch

Maybe this will save someone a bit of pain ...

Download the file above to your home:
`cd`
`wget https://github.com/wpoely86/core/commit/5a49df542697d52bfbde5994ffcfeba88fc8ab9a.patch`
Go to the location of the file to patch:
`cd /var/www/owncloud/lib/private/appframework/http`
Recommended to backup the existing file with:
`cp request.php request.php.old`
Apply the patch:
`patch < ~/5a49df542697d52bfbde5994ffcfeba88fc8ab9a.patch`
Maybe you don't have patch installed, for Ubuntu this is:
`apt-get install patch`

Thanks @wpoely86 and @niobos :)

wpoely86 committed:
Fix remarks from PR

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/8bb776a8-c9be-4d45-8547-71544f71b547) was created.

wpoely86 committed:
Add link to php bug report

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/577a14a0-c3ac-4e98-b4e6-100bcd082707) was created.

wpoely86 commented:
@LukasReschke I agree that it is ugly and the best solution would be to backport the fix in all major distributions. But until that time, something should be done. Apache + php-fpm is not a very exotic setup and user will expect it to work.

I thought there was already an Redhat bug but it's only a CentOS one so I filled one: https://bugzilla.redhat.com/show_bug.cgi?id=1248077

wpoely86 commented on lib/private/appframework/http/request.php:
```diff
@@ -522,12 +522,33 @@ public function getServerProtocol() {
+            $ScriptName = $this->server['SCRIPT_NAME'];
+
+            if (array_key_exists('PATH_INFO', $this->server) === true) {
+                $pos = strpos($this->server['SCRIPT_NAME'], rawurldecode($this->server['PATH_INFO']));
```
@LukasReschke For some reason, `$this->server['PATH_INFO']` does not work...

PVince81 commented:
How to move forward with this PR ?

Xenopathic commented:
Code is OK from my perspective, but there's conflicting views whether this actually works or not - can someone with such a setup verify it actually works?

DeepDiver1975 commented:
@karlitschek @cmonteroluque 

I cannot decide on this change in the current situation -> vote for move to 9.0

karlitschek commented:
yes. unclear and risky. Lets move to 9

DeepDiver1975 commented:
> yes. unclear and risky. Lets move to 9

move applied

PVince81 commented:
> can someone with such a setup verify it actually works?

any update ?

MorrisJobke commented:
@owncloud/qa do we have a RHEL 6/7 or Ubuntu 14.04 with PHP-FPM setup to test this? And verify first if this is still an issue.

wpoely86 commented:
@MorrisJobke It's still an issue with RHEL 7 (don't know about the others).

PVince81 commented:
@owncloud/qa any update ?

VicDeo commented:
Upstream bugreports
https://bugzilla.redhat.com/show_bug.cgi?id=1248077
https://bugs.centos.org/view.php?id=7562

Fixed is released with PHP 5.5.18, PHP 5.6.3 

Looks like a workaround:
https://webtatic.com/packages/php55/
https://webtatic.com/packages/php56/

RealRancor commented:
Follow-up PR which might replace this is in https://github.com/owncloud/core/issues/25800
